assets/src/edit-story/components/postLock/postLock.js

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { useFeatures } from 'flagged';
+import { trackError } from '@web-stories-wp/tracking';
+
+/**
+ * Internal dependencies
+ */
+import { useAPI } from '../../app/api';
+import { useStory } from '../../app/story';
+import { useConfig } from '../../app/config';
+import PostLockDialog from './postLockDialog';
+
+function PostLock() {
+  const {
+    actions: { getStoryLockById, setStoryLockById, deleteStoryLockById },
+  } = useAPI();
+  const {
+    storyId,
+    userId,
+    dashboardLink,
+    postLock: { interval: postLockInterval, showLockedDialog },
+  } = useConfig();
+
+  const { previewLink } = useStory(({ state: { story: { previewLink } } }) => ({
+    previewLink,
+  }));
+  const [showDialog, setShowDialog] = useState(false);
+  const [user, setUser] = useState({});
+  const { enablePostLocking } = useFeatures();
+
+  const closeDialog = useCallback(() => {
+    setShowDialog(false);
+    setStoryLockById(storyId);
+  }, [setStoryLockById, storyId]);
+
+  const doGetStoryLock = useCallback(() => {
+    if (enablePostLocking && showLockedDialog) {
+      getStoryLockById(storyId)
+        .then((result) => {
+          if (result.locked && result.user !== userId) {
+            setShowDialog(true);
+            setUser(result['_embedded'].author[0]);
+          } else {
+            setStoryLockById(storyId);
+          }
+        })
+        .catch((err) => {
+          trackError('post_lock', err.message);
+        });
+    }
+  }, [
+    getStoryLockById,
+    setStoryLockById,
+    setShowDialog,
+    storyId,
+    userId,
+    enablePostLocking,
+    showLockedDialog,
+  ]);
+
+  // Cache it to make it stable in terms of the below timeout
+  const cachedDoGetStoryLock = useRef(doGetStoryLock);
+  useEffect(() => {
+    cachedDoGetStoryLock.current = doGetStoryLock;
+  }, [doGetStoryLock]);
+
+  useEffect(() => {
+    async function releasePostLock() {
+      if (enablePostLocking && showLockedDialog && !showDialog) {
+        await deleteStoryLockById(storyId);
```

Comment:
I think we need to use `navigator.sendBeacon` here, as per https://github.com/google/web-stories-wp/issues/3126#issuecomment-785184082

Reply:
So I found a nice work around this. See docs for `fetch` - https://javascript.info/fetch-api

[KeepAlive](https://javascript.info/fetch-api#keepalive) does what is says it does and is designed for this. 

Also see for background - https://stackoverflow.com/questions/55421459/http-request-beforeunload-sendbeacon-vs-img-src

Reply:
TIL! Very cool.

Sadly, browser support isn't optimal.

Safari and Firefox don't seem to support this yet.

https://bugzilla.mozilla.org/show_bug.cgi?id=1342484
https://bugs.webkit.org/show_bug.cgi?id=168865

Not 100% sure about Safari, as https://wpt.fyi/results/fetch/api/basic/keepalive.html?label=experimental&label=master&aligned says it's supported.

So I think we still need `sendBeacon` to support at Firefox.

Reply:
The issue with using `sendBeacon` is that is not doing to work well with REST API, as it doesn't support sending headers as far as I can see. I will look into it.

Reply:
Fixed in https://github.com/google/web-stories-wp/pull/6497/commits/3f24e3aef30a9989d645a3836b7c608ef32ca372.